File: packages/web/tests/totp.test.ts
import { describe, expect, it } from "vitest";
import {
  decodeBase32,
  formatOtp,
  groupCode,
  InvalidOtpSecret,
  parseOtp,
  secondsRemaining,
  totpAt,
  type OtpConfig,
} from "../src/vault/totp";

/**
 * RFC 6238 Appendix B.
 *
 * The published vectors are the only thing that can actually settle whether
 * this is right. A TOTP implementation that is subtly wrong — off-by-one on the
 * counter, truncating from the wrong offset — still produces six plausible
 * digits, and would fail silently at a sign-in screen with no way for anyone to
 * tell why.
 *
 * The seeds are the RFC's ASCII strings, base32-encoded.
 */
const SEED_SHA1 = "GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ"; // "12345678901234567890"
const SEED_SHA256 = "GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZA====";
const SEED_SHA512 =
  "GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ" +
  "GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNA=";

function config(over: Partial<OtpConfig> = {}): OtpConfig {
  return { secret: SEED_SHA1, digits: 8, period: 30, algorithm: "SHA-1", ...over };
}

describe("RFC 6238 test vectors", () => {
  const cases: [number, OtpConfig, string][] = [
    [59, config(), "94287082"],
    [1111111109, config(), "07081804"],
    [1111111111, config(), "14050471"],
    [1234567890, config(), "89005924"],
    [2000000000, config(), "69279037"],
    [20000000000, config(), "65353130"],

    [59, config({ secret: SEED_SHA256, algorithm: "SHA-256" }), "46119246"],
    [1111111109, config({ secret: SEED_SHA256, algorithm: "SHA-256" }), "68084774"],
    [20000000000, config({ secret: SEED_SHA256, algorithm: "SHA-256" }), "77737706"],

    [59, config({ secret: SEED_SHA512, algorithm: "SHA-512" }), "90693936"],
    [1111111109, config({ secret: SEED_SHA512, algorithm: "SHA-512" }), "25091201"],
    [20000000000, config({ secret: SEED_SHA512, algorithm: "SHA-512" }), "47863826"],
  ];

  for (const [seconds, options, expected] of cases) {
    it(`${options.algorithm} at ${seconds}s produces ${expected}`, async () => {
      expect(await totpAt(options, seconds * 1000)).toBe(expected);
    });
  }

  it("crosses the 32-bit counter boundary correctly", async () => {
    // 20000000000s is past where a naive 32-bit shift silently wraps, which is
    // why the RFC includes it. Worth naming: the failure would appear years
    // from now, not in testing.
    expect(await totpAt(config(), 20000000000 * 1000)).toBe("65353130");
  });
});

describe("six-digit codes, as sites actually issue them", () => {
  it("truncates to the requested number of digits", async () => {
    const six = await totpAt(config({ digits: 6 }), 59_000);
    expect(six).toBe("287082");
    expect(six).toHaveLength(6);
  });

  it("pads a short code rather than dropping a leading zero", async () => {
    // A code of "012345" typed as "12345" is rejected by the site, and the
    // person retyping it has no idea why.
    const padded = String(1).padStart(6, "0");
    expect(padded).toBe("000001");
    const code = await totpAt(config({ digits: 6 }), 1111111109_000);
    expect(code).toMatch(/^\d{6}$/);
  });

  it("holds a code steady across its window and changes at the boundary", async () => {
    const options = config({ digits: 6, period: 30 });
    const atStart = await totpAt(options, 60_000);
    const nearEnd = await totpAt(options, 89_999);
    const after = await totpAt(options, 90_000);
    expect(nearEnd).toBe(atStart);
    expect(after).not.toBe(atStart);
  });

  it("counts down to the moment the code changes", () => {
    const options = config({ period: 30 });
    expect(secondsRemaining(options, 60_000)).toBe(30);
    expect(secondsRemaining(options, 75_000)).toBe(15);
    expect(secondsRemaining(options, 89_000)).toBe(1);
  });
});

describe("taking in a key the way a site hands it over", () => {
  it("accepts a bare secret, spaces and all", () => {
    const parsed = parseOtp("gezd gnbv gy3t qojq gezd gnbv gy3t qojq");
    expect(parsed.secret).toBe(SEED_SHA1);
    expect(parsed.digits).toBe(6);
    expect(parsed.period).toBe(30);
  });

  it("accepts the otpauth URI behind a QR code", () => {
    const parsed = parseOtp(
      `otpauth://totp/GitHub:maria%40example.com?secret=${SEED_SHA1}&issuer=GitHub&digits=6&period=30`,
    );
    expect(parsed.secret).toBe(SEED_SHA1);
    expect(parsed.issuer).toBe("GitHub");
    expect(parsed.label).toBe("maria@example.com");
  });

  it("carries a non-standard period and digit count", () => {
    const parsed = parseOtp(
      `otpauth://totp/Bank?secret=${SEED_SHA1}&digits=8&period=60&algorithm=SHA256`,
    );
    expect(parsed.digits).toBe(8);
    expect(parsed.period).toBe(60);
    expect(parsed.algorithm).toBe("SHA-256");
  });

  it("round-trips through the single stored field", () => {
    const original = parseOtp(
      `otpauth://totp/Bank:me?secret=${SEED_SHA1}&digits=8&period=60&algorithm=SHA256&issuer=Bank`,
    );
    const again = parseOtp(formatOtp(original));
    expect(again.secret).toBe(original.secret);
    expect(again.digits).toBe(original.digits);
    expect(again.period).toBe(original.period);
    expect(again.algorithm).toBe(original.algorithm);
  });

  it("rejects a mistyped key rather than producing codes nothing accepts", () => {
    // Silently accepting this is the cruel failure: the app shows confident
    // six-digit codes and the site refuses every one of them.
    expect(() => parseOtp("not a real key!!")).toThrow(InvalidOtpSecret);
    expect(() => parseOtp("")).toThrow(InvalidOtpSecret);
    expect(() => decodeBase32("0189")).toThrow(InvalidOtpSecret);
  });

  it("says plainly when a key is counter-based rather than time-based", () => {
    expect(() => parseOtp(`otpauth://hotp/Bank?secret=${SEED_SHA1}&counter=1`)).toThrow(
      /counter-based/,
    );
  });
});

describe("reading a code out loud", () => {
  it("splits into halves", () => {
    expect(groupCode("123456")).toBe("123 456");
    expect(groupCode("12345678")).toBe("1234 5678");
  });
});

File: packages/android/vault/src/test/kotlin/app/keyweb/vault/TotpTest.kt
package app.keyweb.vault

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

/**
 * RFC 6238 Appendix B, and the same vectors the TypeScript suite uses.
 *
 * Two implementations of TOTP that agree with each other but not with the RFC
 * would be consistently wrong, which is worse than one being broken: every code
 * would be rejected everywhere, with no clue as to why.
 */
class TotpTest {

    private val sha1 = "GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ"
    private val sha256 = "GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZA===="
    private val sha512 =
        "GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ" +
            "GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNA="

    private fun config(secret: String = sha1, algorithm: String = "SHA1") =
        OtpConfig(secret = secret, digits = 8, period = 30, algorithm = algorithm)

    @Test
    fun `matches the published vectors`() {
        val cases = listOf(
            Triple(59L, config(), "94287082"),
            Triple(1111111109L, config(), "07081804"),
            Triple(1111111111L, config(), "14050471"),
            Triple(1234567890L, config(), "89005924"),
            Triple(2000000000L, config(), "69279037"),
            // Past where a 32-bit counter silently wraps. The failure this
            // catches would otherwise appear years from now.
            Triple(20000000000L, config(), "65353130"),

            Triple(59L, config(sha256, "SHA256"), "46119246"),
            Triple(1111111109L, config(sha256, "SHA256"), "68084774"),
            Triple(20000000000L, config(sha256, "SHA256"), "77737706"),

            Triple(59L, config(sha512, "SHA512"), "90693936"),
            Triple(1111111109L, config(sha512, "SHA512"), "25091201"),
            Triple(20000000000L, config(sha512, "SHA512"), "47863826"),
        )
        for ((seconds, options, expected) in cases) {
            assertEquals(expected, Totp.at(options, seconds * 1000), "at ${seconds}s")
        }
    }

    @Test
    fun `produces the six digits sites actually ask for`() {
        assertEquals("287082", Totp.at(config().copy(digits = 6), 59_000))
    }

    @Test
    fun `holds steady across a window and changes at the boundary`() {
        val options = config().copy(digits = 6)
        assertEquals(Totp.at(options, 60_000), Totp.at(options, 89_999))
        assertTrue(Totp.at(options, 90_000) != Totp.at(options, 60_000))
    }

    @Test
    fun `counts down to the moment the code changes`() {
        assertEquals(30, Totp.secondsRemaining(config(), 60_000))
        assertEquals(15, Totp.secondsRemaining(config(), 75_000))
        assertEquals(1, Totp.secondsRemaining(config(), 89_000))
    }

    @Test
    fun `takes a bare secret with the spaces a site prints`() {
        val parsed = Totp.parse("gezd gnbv gy3t qojq gezd gnbv gy3t qojq")
        assertEquals(sha1, parsed.secret)
        assertEquals(6, parsed.digits)
    }

    @Test
    fun `takes the otpauth URI behind a QR code`() {
        val parsed = Totp.parse(
            "otpauth://totp/GitHub:maria%40example.com?secret=$sha1&issuer=GitHub&digits=6",
        )
        assertEquals(sha1, parsed.secret)
        assertEquals("GitHub", parsed.issuer)
        assertEquals("maria@example.com", parsed.label)
    }

    @Test
    fun `round-trips through the single stored field`() {
        val original = Totp.parse(
            "otpauth://totp/Bank:me?secret=$sha1&digits=8&period=60&algorithm=SHA256&issuer=Bank",
        )
        val again = Totp.parse(Totp.format(original))
        assertEquals(original.secret, again.secret)
        assertEquals(original.digits, again.digits)
        assertEquals(original.period, again.period)
        assertEquals(original.algorithm, again.algorithm)
    }

    @Test
    fun `refuses a mistyped key rather than printing codes nothing accepts`() {
        // The cruel failure: confident six-digit codes that every site rejects.
        assertFailsWith<InvalidOtpSecret> { Totp.parse("not a real key!!") }
        assertFailsWith<InvalidOtpSecret> { Totp.parse("") }
        assertFailsWith<InvalidOtpSecret> { Totp.decodeBase32("0189") }
    }

    @Test
    fun `says plainly when a key is counter-based`() {
        val failure = assertFailsWith<InvalidOtpSecret> {
            Totp.parse("otpauth://hotp/Bank?secret=$sha1&counter=1")
        }
        assertTrue(failure.message!!.contains("counter-based"))
    }

    @Test
    fun `groups a code for reading aloud`() {
        assertEquals("123 456", Totp.group("123456"))
        assertEquals("1234 5678", Totp.group("12345678"))
    }
}
